Pattern properties support

delvedor opened the pull request:
Added support for `patternProperties`.
See discussion in #5 .

mcollina commented on README.md:
```diff
@@ -119,6 +120,74 @@ const obj = {
+    },
+    'num': {
+      type: 'number'
+    }
```
We should support regexps:

```js
    "patternProperties": {
        "^(/[^/]+)+$": {}
    },
```

delvedor replied:
I'll update this with a more complex example.

mcollina commented on index.js:
```diff
@@ -131,11 +132,66 @@ function $asRegExp (reg) {
+  } else {
+    throw new Error('Cannot coerce ' + value + ' to ' + type)
+  }
+}
```
why are you doing a single function, when `type` is known at startup time?

mcollina commented on test.js:
```diff
@@ -316,3 +316,112 @@ test('missing values', (t) => {
+  })
+
+  obj = { foo: 'true', ofoo: 0, arrfoo: [1, 2], objfoo: { tyrion: 'lannister' } }
+  t.equal('{"foo":[],"ofoo":[],"arrfoo":[],"objfoo":[]}', stringify(obj))
```
You can split each of those tests in its own `test`. They are hard to read.
You can also split each feature in its own file.
See https://github.com/mcollina/pino as an example.

delvedor replied:
Yeah, you are right, I'll update it.

mcollina commented on README.md:
```diff
@@ -119,6 +120,74 @@ const obj = {
+      type: 'number'
+    }
+  }
+})
```
I don't think this is correct, it is missing the  `obj` property.

mcollina commented on README.md:
````diff
@@ -119,6 +120,74 @@ const obj = {
+console.log(stringify(obj)) // '{"nickname":"nick","matchfoo":"42","otherfoo":"str","matchnum":3}'
+```
+`Object` and `Array` will be stringified as `{}` and `[]`.  
+Since `fast-json-stringify` is faster than `JSON.stringify` because works on a **fixed schema**, we strongly advice you to not pass complex objects or array as object values, but decompose them in their individual properties and stringify them.  
````
to not pass -> not to pass

mcollina replied:
Also, this sentence is probably not ok. It's perfectly fine to pass complex objects, as long as the schema stays fixed.

delvedor replied:
I'll remove all this part, since in the next pr I'll fix the nested object properties, as they works now in case of the array.

delvedor commented:
I've updated the docs and fixed the nested array/objects stringify.
In the near future we must add to the docs an explanation of why `fast-json-stringify` needs a fixed schema to stringify properties.

I think could be a good idea throw if object or array has not properties/items fields.

mcollina commented:
Great job :).

What happens if one of the properties defined in `properties`  is also matched by `patternProperties`?

delvedor commented:
`patternProperties` does not overwrite `properties`, you can check it [here](https://github.com/mcollina/fast-json-stringify/blob/0fd2d0c42271d58c2462dfb9ac2fa6bd54eeba01/index.js#L138) and [here](https://github.com/mcollina/fast-json-stringify/blob/0fd2d0c42271d58c2462dfb9ac2fa6bd54eeba01/test.js#L341-L360) :)

mcollina commented on index.js:
```diff
@@ -4,7 +4,7 @@ function build (schema) {
   var code = `
     'use strict'
-
+    const properties = ${JSON.stringify(schema.properties)}
```
Nice one, I completely missed this. Can you please explicit why this is needed with a comment?
